File: DP_Problems/canSum/canSum.js
const canSum = (targetSum, numberArray, memo = {}) => {
  if (targetSum in memo) return memo[targetSum]
  if (targetSum === 0) return true
  if (targetSum < 0) return false

  for (let n of numberArray) {
    const remainder = targetSum - n
    if (canSum(remainder, numberArray, memo) === true) {
      memo[targetSum] = true
      return true
    }
  }

  memo[targetSum] = false
  return false
}

console.log(canSum(7, [2, 3]))
console.log(canSum(7, [5, 3, 4, 7]))
console.log(canSum(7, [2, 4]))
console.log(canSum(8, [2, 3, 5]))
console.log(canSum(300, [7, 14]))

File: DP_Problems/canSum/canSum.cpp
#include <iostream>
#include <vector>

using namespace std;

bool canSum(int targetSum, vector<int> numbersArray)
{
    if (targetSum == 0)
        return true;
    if (targetSum < 0)
        return false;

    for (int i = 0; i < numbersArray.size(); i++)
    {
        int remainder = targetSum - numbersArray[i];
        if (canSum(remainder, numbersArray) == true)
            return true;
    }

    return false;
}

int main()
{
    vector<int> temp{2, 3};
    cout << canSum(7, temp) << endl;
    temp = {5, 3, 4, 7};
    cout << canSum(7, temp) << endl;
    temp = {2, 4};
    cout << (canSum(7, temp)) << endl;

    return 0;
}